lib/mix/test/mix/tasks/deps.git_test.exs

```diff
@@ -478,6 +478,204 @@ defmodule Mix.Tasks.DepsGitTest do
+        message = "* Getting git_repo (#{fixture_path("git_repo")} - main)"
+        assert_received {:mix_shell, :info, [^message]}
+        assert_shallow("deps/git_repo", 1)
+      end)
+    end
+
+    test "can be used with ref" do
+      [last, _ | _] = get_git_repo_revs("git_repo")
+
+      Process.put(:git_repo_opts, depth: 1, ref: last)
+
+      in_fixture("no_mixfile", fn ->
+        Mix.Project.push(GitApp)
+
+        Mix.Tasks.Deps.Get.run([])
+        message = "* Getting git_repo (#{fixture_path("git_repo")} - #{last})"
+        assert_received {:mix_shell, :info, [^message]}
+        assert_shallow("deps/git_repo", 1)
+      end)
+    end
+
+    test "changing refspec updates retaining depth" do
+      [last, first | _] = get_git_repo_revs("git_repo")
+
+      Process.put(:git_repo_opts, ref: first, depth: 1)
+
+      in_fixture("no_mixfile", fn ->
+        Mix.Project.push(GitApp)
+
+        Mix.Tasks.Deps.Get.run([])
+        message = "* Getting git_repo (#{fixture_path("git_repo")} - #{first})"
+        assert_received {:mix_shell, :info, [^message]}
+        assert_shallow("deps/git_repo", 1)
+        assert File.read!("mix.lock") =~ first
+
+        # Change refspec
+        update_dep(ref: last, depth: 1)
+        Mix.Tasks.Deps.Get.run([])
+        assert_shallow("deps/git_repo", 1)
+        assert File.read!("mix.lock") =~ last
+      end)
+    end
+
+    test "removing depth retains shallow repository" do
+      # The expectation is that one needs to deps.clean and deps.get the
+      # dependency for a full clone.
+      Process.put(:git_repo_opts, depth: 1)
+
+      in_fixture("no_mixfile", fn ->
+        Mix.Project.push(GitApp)
+
+        Mix.Tasks.Deps.Get.run([])
+        message = "* Getting git_repo (#{fixture_path("git_repo")})"
+        assert_received {:mix_shell, :info, [^message]}
+        assert_shallow("deps/git_repo", 1)
+
+        # Remove depth
+        update_dep([])
+        Mix.Tasks.Deps.Get.run([])
+        refute File.read!("mix.lock") =~ "depth:"
+        assert File.exists?("deps/git_repo/.git/shallow")
+
+        assert System.cmd("git", ~w[--git-dir=deps/git_repo/.git rev-list --count HEAD]) ==
+                 {"1\n", 0}
+      end)
+    end
+
+    @tag :git_sparse
+    test "can be used with sparse checkout" do
+      Process.put(:git_repo_opts, sparse: "sparse_dir", depth: 1)
+
+      in_fixture("no_mixfile", fn ->
+        Mix.Project.push(GitApp)
+
+        Mix.Tasks.Deps.Get.run([])
+        message = "* Getting git_repo (#{fixture_path("git_repo")})"
+        assert_received {:mix_shell, :info, [^message]}
+        assert_shallow("deps/git_repo", 1)
+
+        refute File.exists?("deps/git_repo/mix.exs")
+        assert File.exists?("deps/git_repo/sparse_dir/mix.exs")
+        assert File.read!("mix.lock") =~ "sparse: \"sparse_dir\""
+      end)
+    end
+
+    test "can be used with subdir" do
+      Process.put(:git_repo_opts, subdir: "sparse_dir", depth: 1)
```

Comment:
I noticed that the `:subdir` options forces the creation of the subdir, even if it doesn't exist in the repository. It was curious to me that the test fixtures don't create a subdir named "subdir" (what the existing subdir test uses), and so used the "sparse_dir" path here, and also asserting it contains what we expect (unlike the existing subdir test).